sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEPropagateEmptyRelation.scala

```diff
@@ -53,8 +54,11 @@ object AQEPropagateEmptyRelation extends PropagateEmptyRelationBase {
       empty(j)
   }
 
-  // TODO we need use transformUpWithPruning instead of transformUp
-  def apply(plan: LogicalPlan): LogicalPlan = plan.transformUp {
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformUpWithPruning(
+    // LOCAL_RELATION and TRUE_OR_FALSE_LITERAL pattern are matched at `PropagateEmptyRelationBase`
+    // LOGICAL_QUERY_STAGE pattern is matched at `AQEPropagateEmptyRelation`
+    // We can not specify ruleId here since the LogicalQueryStage is not immutable.
+    _.containsAnyPattern(LOGICAL_QUERY_STAGE, LOCAL_RELATION, TRUE_OR_FALSE_LITERAL)) {
```

Comment:
This looks correct, but may not have good pruning power?  How about:

t = > t.containsAllPatterns(LOGICAL_QUERY_STAGE, JOIN, AND_OR, NULL_CHECK)
      || t.containsAnyPatterns(LOCAL_RELATION, TRUE_OR_FALSE_LITERAL)

The first condition covers eliminateSingleColumnNullAwareAntiJoin while the second covers commonApplyFunc.